### PR TITLE
New Device Flyout - Basic 

### PR DESCRIPTION
**Pull request by mechaffin**
# Type of change? 

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

Implements the New Device flyout with very basic functionality. Only a single physical device with default options can be provisioned. Context menu is hooked up.

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: public/locales/en/translations.json [hunk 1]
```diff
@@ -115,12 +115,33 @@
...
+      "summaryHeader": "Summary",
       "affected": "Affected devices",
       "cancel": "Cancel",
       "apply": "Delete",
+      "pending": "Deleting devices... ",
       "applySuccess": "devices deleted successfully!",
       "close": "Close"
     }
+    ,
+    "new": {
```
Comment (spryor) [inline public/locales/en/translations.json]: I think these names need to be more explicit for readability purposes. Something like `flyoutNewDevice` or wrap them in a `flyouts` container.

File: src/components/pages/devices/devices.js [hunk 2]
```diff
@@ -37,14 +38,16 @@ export class Devices extends Component {
 
   closeFlyout = () => this.setState(closedFlyoutState);
 
+  openNewDeviceFlyout = () => this.setState({ openFlyoutName: "new-device" });
+
   onSoftSelectChange = ({ id }) => this.setState({
-    flyoutOpen: true,
+    openFlyoutName: "details",
```
Comment (spryor) [inline src/components/pages/devices/devices.js]: single quotes

File: src/components/pages/devices/devices.js [hunk 3]
```diff
@@ -59,10 +62,13 @@ export class Devices extends Component {
       getSoftSelectId: this.getSoftSelectId,
       t: this.props.t
     };
+    const detailsFlyoutOpen = this.state.openFlyoutName === "details";
```
Comment (spryor) [inline src/components/pages/devices/devices.js]: single quotes on these two lines

File: src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js [hunk 4]
```diff
@@ -86,7 +89,8 @@ export class DeviceDelete extends Component {
     } = this.state;
 
     const summaryCount = changesApplied ? successCount : physicalDevices.length;
-    const summaryMessage = changesApplied ? t('devices.delete.applySuccess') : t('devices.delete.affected');
+    const summaryMessage = isPending ? t('devices.delete.pending') : changesApplied ? t('devices.delete.applySuccess') : t('devices.delete.affected');
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDelete/deviceDelete.js]: I wouldn't nest ternary operators like this since it is hard to read.

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.container.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+import { connect } from 'react-redux';
+import { translate } from 'react-i18next';
+import { DeviceNew } from './deviceNew';
+import { redux as devicesRedux } from 'store/reducers/devicesReducer';
+
+// Pass the device details
+const mapStateToProps = state => null;
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.container.js]: Pass `null` into the connect method and remove this line.

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  </FormSection>
+);
+
+const DeviceConnectionString = ({ label, deviceId, hostName, sharedAccessKey }) => (
+  <DeviceDetail label={label} value={`HostName=${hostName};DeviceId=${deviceId};SharedAccessKey=${sharedAccessKey}`} />
+);
+
+
+const ProvisionedDevice = ({ device, t }) => {
+  const { id, ioTHubHostName: hostName, authentication: { primaryKey }, authentication: { secondaryKey } } = device;
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.js]: break into multiple lines

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+  return (
+    <div>
+      <pre>{JSON.stringify(device, null, 2)}</pre>
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.js]: Why is this here? It is a left over?

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  componentWillUnmount() {
+    if (this.subscription) this.subscription.unsubscribe();
+  }
+
+  provisionDevices = () => {
+    this.setState({ isPending: true });
+
+    console.log(this.state);
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.js]: Remove console logs

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    console.log(this.countLink);
+
+    //TODO: Create the form to enable user input.
+    this.subscription = IoTHubManagerService.provisionDevice({ "Id": "", "Authentication": {}, "IsSimulated": false })
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.js]: Just like we have models that map from service response to UI model, we want to have models that from to request objects. You need to make on here. Also use single quotes.

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.js [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+      .subscribe(
+        provisionedDevice => {
+          this.setState({ provisionedDevice, successCount: this.state.formData.count, isPending: false, changesApplied: true });
+          return this.props.insertDevice(provisionedDevice);
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.js]: Remove the return.

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.scss [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    width: 100%;
+
+    .section-header { font-size: inherit; }
+    .device-detail-contents {
+      display: flex;
+      align-items: flex-start;
+      word-wrap: break-word;
+      flex-grow: 0;
+
+      .device-detail-value { flex-basis: 90%; }
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.scss]: new line

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.scss [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+  .summary-icon svg {
+    @include square-px-rem(16px);
+    @include rem-fallback(margin-left, 8px);
+  }
+
+  .device-detail {
+    width: 100%;
+
+    .section-header { font-size: inherit; }
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceNew/deviceNew.scss]: newline

File: src/components/shared/forms/styles/summarySection.scss [hunk 13]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+@import 'src/styles/mixins';
+@import 'src/styles/themes';
+
+.summary-section {
+  @include rem-fallback(margin, 12px 0px);
+
+  .section-header {
```
Comment (spryor) [inline src/components/shared/forms/styles/summarySection.scss]: one line

File: src/components/shared/forms/summarySection.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import { joinClasses } from 'utilities';
+import {
```
Comment (spryor) [inline src/components/shared/forms/summarySection.js]: one line

File: src/services/models/iotHubManagerModels.js [hunk 15]
```diff
@@ -7,7 +7,9 @@ import { reshape } from 'utilities';
 // TODO: Map to backend models and add links to github
 
 export const toDevicesModel = (response = {}) => (response.items || [])
-  .map((device = {}) => reshape(device, {
+  .map((device = {}) => toDeviceModel(device));
```
Comment (spryor) [inline src/services/models/iotHubManagerModels.js]: `.map(toDeviceModel);`

File: public/locales/en/translations.json (hunk 1 above)
Comment (mechaffin) [reply public/locales/en/translations.json]: That would be a change for all of the flyout translations, not just "new".  So let's do that in another changeset.

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.js (hunk 7 above)
Comment (mechaffin) [reply src/components/pages/devices/flyouts/deviceNew/deviceNew.js]: Because this page is not done yet... and so yes, its temporary. The next version of this branch won't have it.

File: src/components/pages/devices/flyouts/deviceNew/deviceNew.js (hunk 9 above)
Comment (mechaffin) [reply src/components/pages/devices/flyouts/deviceNew/deviceNew.js]: Yes, this is temporary... since this is just a super basic implementation for now.

**Comment by mechaffin** [comment]
Closing this PR to create a new one... now that the branch has been rebased to archMigration.
